group/assignment6/move_cars.py

```diff
@@ -0,0 +1,59 @@
+def generate_moves(start, final):
+    """ ith element of start start is the car currently in the ith space
+    ith element of final start is the car that should be in ith space
+    empty space is 0
+    yields optimal sequence of moves in format (car number, old space, new space) """
+
+    # validation of the input
+    start_set = set(start)
+    final_set = set(final)
+    #print(start_set, final_set)
+    if len(start_set) != len(start) or start_set != final_set or \
+       len(final_set) != len(final) or \
+       start_set != final_set and 0 not in start_set:
+        yield None
+
+    # pos will contain car positions in another form
+    pos = [0] * len(start)
+    if 0 in start_set:
+        for place, car in enumerate(start):
+            pos[car] = place
+
+        # main loop
+        for car, place in enumerate(pos):
+            if final[place] == car:
+                continue
+
+            if car:  # this is a cycle fix
+                yield car, place, pos[0]
+                pos[0], pos[car] = pos[car], pos[0]
+
+            car = final[pos[0]]
+            while car:
+                yield car, pos[car], pos[0]
+                pos[0], pos[car] = pos[car], pos[0]
+                car = final[pos[0]]
+
+def fancy_output(start, final):
+    result = generate_moves(start, final)
+    for i in result:
+        if i:
+            print("Move car number {} from {} to {}".format(*i))
+        else:
+            print("Unsolvable")
```

Comment:
Unsolvable?
More like "Wrong input format"

Opt3 will actually have unsolvable cases (and I'll try to get around to writing it in the next couple of days), so best to keep the distinction

Maybe we want exceptions for wrong input format? Possibly checking the Google code style beforehand

Reply:
Yes, I think that checking Google code style is a good idea. I will do it and write a conclusion about suitable treatment of a wrong input etc in comments.
I also think that we should discuss this moment with other group members, as a part of program design.